### PR TITLE
[Feature][Json-format] support read format for pulsar

### PR DESCRIPTION
**Pull request by laglangyue**
## Purpose of this pull request

#4066

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by laglangyue** [comment]
Currently, the format plugin is instantiated in the connector. Will it be used to instantiate the format plugin through DeserializationFormatFactory in the future

**Comment by hailin0** [comment]
Please add e2e

**Comment by laglangyue** [comment]
> Please add e2e

ok，and this is right?


> Currently, the format plugin is instantiated in the connector. Will it be used to instantiate the format plugin through DeserializationFormatFactory in the future

**Comment by hailin0** [comment]
> 



> > Please add e2e
> 
> ok，and this is right?
> 
> > Currently, the format plugin is instantiated in the connector. Will it be used to instantiate the format plugin through DeserializationFormatFactory in the future

👌

**Comment by liugddx** [comment]
Please fix ci error.

**Comment by laglangyue** [comment]
image


I find a bug,sometimes,the split thead would exit but not notify source,I will try to fix it

**Comment by laglangyue** [comment]
image
the running is always false,I fix it.

image
image
the notifyCheckpointComplete is not be invoked，therefore,the task can't be complted，I don't know why? and I try to fix it

**Comment by laglangyue** [comment]
in pulsar  topic ,the data is 

image


image


and the data is not one by one,the pulsar-canal-connector (In pulsar) will aggregate some message.
should we split?

**Comment by laglangyue** [comment]
> in pulsar topic ,the data is
> 
> image
> 
> image
> 
> and the data is not one by one,the pulsar-canal-connector (In pulsar) will aggregate some message. should we split?

@hailin0  what do you think?

**Comment by laglangyue** [comment]
I has finished the e2e,Please review it.
@hailin0

**Comment by laglangyue** [comment]
and the canal_json of Pulsar-canal-connector is 
```json
{
"id":"",
"timestamp":"",
"message":"[canal json list]"
}
```
so I decorator the Canal format by the class named PulsarCanalDecorator

File: seatunnel-connectors-v2/connector-pulsar/pom.xml [hunk 1]
```diff
@@ -100,5 +100,23 @@
...
+            <groupId>org.testcontainers</groupId>
+            <artifactId>testcontainers</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>pulsar</artifactId>
+            <version>${testcontainer.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-pulsar/pom.xml]: move to e2e module

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/reader/PulsarSourceReader.java [hunk 2]
```diff
@@ -203,6 +199,10 @@ protected PulsarSplitReaderThread createPulsarSplitReaderThread(PulsarPartitionS
     public void handleNoMoreElements(String splitId, MessageId messageId) {
         LOG.info("Reader received the split {} NoMoreElements event.", splitId);
         pendingCursorsToFinish.put(splitId, messageId);
+        // BOUNDED not trigger snapshot and notifyCheckpointComplete
+        if (context.getBoundedness() == Boundedness.BOUNDED) {
+            finishedSplits.add(splitId);
+        }
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/reader/PulsarSourceReader.java:205]: good

File: docs/en/connector-v2/source/Pulsar.md [hunk 3]
```diff
@@ -141,7 +150,7 @@ source {
   Pulsar {
   	topic = "example"
   	subscription.name = "seatunnel"
-    client.service-url = "localhost:pulsar://localhost:6650"
+    client.service-url = "pulsar://localhost:6650"
```
Comment (hailin0) [inline docs/en/connector-v2/source/Pulsar.md:148]: uodate `Changelog`

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/CanalToPulsarIT.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+    @Override
+    public void tearDown() {
+        MYSQL_CONTAINER.close();
+        PULSAR_CONTAINER.close();
+        CANAL_CONTAINER.close();
+    }
+
+    @TestTemplate
+    void testContainer(TestContainer container) throws IOException, InterruptedException {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/CanalToPulsarIT.java]:
```suggestion
    void testCanalFormatMessages(TestContainer container) throws IOException, InterruptedException {
```

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/CanalToPulsarIT.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        Container.ExecResult execResult = container.executeJob("/cdc_canal_pulsar_to_console.conf");
+        Assertions.assertEquals(0, execResult.getExitCode(), execResult.getStderr());
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/CanalToPulsarIT.java:304]: reference check cdc sink output

https://github.com/apache/incubator-seatunnel/pull/3950#discussion_r1090141130

https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/CanalToKafkaIT.java#L265

**Review by hailin0 — COMMENTED** [review]
update `release-note.md`

File: seatunnel-connectors-v2/connector-pulsar/pom.xml (hunk 1 above)
Comment (laglangyue) [reply seatunnel-connectors-v2/connector-pulsar/pom.xml]: Connector-v2-release-state.md

> update `release-note.md`

you mean Connector-v2-release-state.md?

**Comment by hailin0** [comment]
> update `release-note.md`

here
https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md

**Comment by laglangyue** [comment]
> > update `release-note.md`
> 
> here https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md

done

**Comment by laglangyue** [comment]
PTAL
@hailin0 @liugddx @TyrantLucifer

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/CanalToPulsarIT.java (hunk 5 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/CanalToPulsarIT.java:304]: done

**Review by hailin0 — DISMISSED** [review]

**Comment by laglangyue** [comment]
I find a new problem with this, in ci environment, the pulsar container standalone con't start in 1 core CPU VM.I will fix it soon.

**Comment by laglangyue** [comment]
spark3 container didn't has the command 'curl'
so i add it
``` java
        // ensure install curl
        master.execInContainer("bash", "-c" ,"apt-get update -y && apt-get install curl");
```

**Comment by laglangyue** [comment]
image
I find kafka e2e failed

**Comment by laglangyue** [comment]
@hailin0 @liugddx @TyrantLucifer 
there some ci timeout, but it is not caused by my code in my opinion.
should split more part for e2e?

**Review by hailin0 — DISMISSED** [review]

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/JsonDeserializationSchema.java [hunk 6]
```diff
@@ -59,6 +61,7 @@ public class JsonDeserializationSchema implements DeserializationSchema<SeaTunne
     /** Object mapper for parsing the JSON. */
     private final ObjectMapper objectMapper = new ObjectMapper();
 
+    @Builder
```
Comment (TyrantLucifer) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/JsonDeserializationSchema.java]: Why add this?

File: seatunnel-formats/seatunnel-format-text/src/main/java/org/apache/seatunnel/format/text/TextFormatFactory.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (TyrantLucifer) [inline seatunnel-formats/seatunnel-format-text/src/main/java/org/apache/seatunnel/format/text/TextFormatFactory.java]: The correct approach is to put all changes related to this PR together and merge them into a single commit. Any changes unrelated to this PR should be placed in another PR. This helps keep the purpose and scope of each PR clear, and facilitates code review and management. If there are changes unrelated to this PR, they can be mentioned in the commit message and placed in another PR.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java [hunk 8]
```diff
@@ -103,4 +106,19 @@ private static String formatUnderScoreCase(String camel) {
         }
         return underScore.toString();
     }
+
+    public static <T> T get(
```
Comment (TyrantLucifer) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java]: Why add this?

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java [hunk 9]
```diff
@@ -244,6 +244,15 @@ public static ObjectNode parseObject(String text) {
         }
     }
 
+    public static ObjectNode parseObject(byte[] content) {
```
Comment (TyrantLucifer) [inline seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java:239]: return parseObject(String.valueOf(content)), or change parseObject(String json) to return parseObject(json.getBytes());

**Review by TyrantLucifer — CHANGES_REQUESTED** [review]

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java (hunk 8 above)
Comment (laglangyue) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java:110]: it is Overload method

just like this

image

Comment (laglangyue) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java]: it's a Overload method,maybe I can convert Config object to Map，and use the original method

Comment (laglangyue) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java]: I think the method overload is necessary.
image

There are many config objects in our project

image

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java (hunk 9 above)
Comment (laglangyue) [reply seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java:239]: you mean thart use new String(byte[]).
we use parseObject() to convert ST-Row to JsonObject， Jackson can directly read the byte array.
For big josn, read byte array directly is better.

and new String() may cause additional performance loss in my opinion.

so I the overload is necessary

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/JsonDeserializationSchema.java (hunk 6 above)
Comment (laglangyue) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/JsonDeserializationSchema.java]: yes, the parameters are not many,I remove it.

File: seatunnel-formats/seatunnel-format-text/src/main/java/org/apache/seatunnel/format/text/TextFormatFactory.java (hunk 7 above)
Comment (laglangyue) [reply seatunnel-formats/seatunnel-format-text/src/main/java/org/apache/seatunnel/format/text/TextFormatFactory.java]: At the beginning, there were some bugs in the pulsar connector, and the batch task bug fix for the text format was completed first.
I will split this.

**Comment by laglangyue** [comment]
@TyrantLucifer PATL again

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java (hunk 8 above)
Comment (TyrantLucifer) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionUtil.java]: cc @ashulin PTAL

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java (hunk 9 above)
Comment (TyrantLucifer) [reply seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java:239]: It's does not make sense, you can try to change the logic of parseObject(String json)

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/canal/CanalJsonDeserializationSchema.java [hunk 10]
```diff
@@ -137,7 +141,7 @@ public void deserialize(byte[] message, Collector<SeaTunnelRow> out) {
             }
             throw new SeaTunnelJsonFormatException(
                     CommonErrorCode.JSON_OPERATION_FAILED,
-                    format("Null data value \"%s\" Cannot send downstream", new String(message)));
+                    format("Null data value \"%s\" Cannot send downstream", jsonNode.asText()));
```
Comment (TyrantLucifer) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/canal/CanalJsonDeserializationSchema.java]: @MonsterChenzhuo PTAL

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java [hunk 11]
```diff
@@ -233,15 +235,8 @@ public static String toJsonString(Object object) {
...
-        try {
-            if (text.isEmpty()) {
-                return parseObject(text, ObjectNode.class);
-            } else {
-                return (ObjectNode) OBJECT_MAPPER.readTree(text);
-            }
-        } catch (Exception e) {
-            throw new RuntimeException("String json deserialization exception.", e);
-        }
+        assert text != null;
```
Comment (TyrantLucifer) [inline seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java]: Do not use `assert` in codes, it only used in test cases

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/canal/CanalJsonDeserializationSchema.java (hunk 10 above)
Comment (laglangyue) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/canal/CanalJsonDeserializationSchema.java]: image
I fix this

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java (hunk 11 above)
Comment (laglangyue) [reply seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/JsonUtils.java]: done

**Comment by laglangyue** [comment]
@hailin0 @TyrantLucifer PTAL again

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java [hunk 12]
```diff
@@ -70,6 +70,8 @@ public void startUp() throws Exception {
         // cluster mode, we need to
         // start a worker.
         Startables.deepStart(Stream.of(master)).join();
+        // ensure install curl
+        master.execInContainer("bash", "-c", "apt-get update -y && apt-get install curl");
```
Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java]: revert, the newest container has been resolved this problem,

File: seatunnel-examples/seatunnel-spark-connector-v2-example/pom.xml [hunk 13]
```diff
@@ -107,6 +107,19 @@
             <artifactId>lz4</artifactId>
             <version>1.3.0</version>
         </dependency>
+
+        <dependency>
```
Comment (TyrantLucifer) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/pom.xml]: revert

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/batch_pulsar_to_console.conf [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    admin.service-url = "http://pulsar.batch.e2e:8080"
+    subscription.name = "e2e"
+    topic = "topic-it"
+    cursor.startup.mode = "EARLIEST"
+    cursor.stop.mode = "LATEST"
+    format = json
+    result_table_name = "pulsar_canal"
+    schema = {
+      fields {
+        uuid = string
```
Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/batch_pulsar_to_console.conf]: test all data types

Comment (laglangyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/batch_pulsar_to_console.conf]: done

File: seatunnel-examples/seatunnel-spark-connector-v2-example/pom.xml (hunk 13 above)
Comment (laglangyue) [reply seatunnel-examples/seatunnel-spark-connector-v2-example/pom.xml]: done

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java (hunk 12 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java]: done

**Review by hailin0 — DISMISSED** [review]
LGTM

**Review by EricJoy2048 — DISMISSED** [review]

**Review by EricJoy2048 — APPROVED** [review]

**Review by hailin0 — APPROVED** [review]

**Review by TyrantLucifer — APPROVED** [review]

**Review by TyrantLucifer — APPROVED** [review]
